packages/core/contracts/l2/polygon/OracleRootTunnel.sol

```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "./tunnel/FxBaseRootTunnel.sol";
+import "./OracleBaseTunnel.sol";
+import "../../oracle/interfaces/OracleAncillaryInterface.sol";
+
+/**
+ * @title Adapter deployed on mainnet that validates and sends price requests from sidechain to the DVM on mainnet.
+ * @dev This contract must be a registered financial contract in order to make DVM price requests.
+ */
+contract OracleRootTunnel is OracleBaseTunnel, FxBaseRootTunnel {
+    constructor(
+        address _checkpointManager,
+        address _fxRoot,
+        address _finderAddress
+    ) OracleBaseTunnel(_finderAddress) FxBaseRootTunnel(_checkpointManager, _fxRoot) {}
+
+    /**
+     * @notice This is the first method that should be called in order to publish a price request to the sidechain.
+     * @dev Publishes the DVM resolved price for the price request, or reverts if not resolved yet. 
+     * @param identifier Identifier of price request to resolve.
+     * @param time Timestamp of price request to resolve.
+     * @param ancillaryData extra data of price request to resolve.
+     */
+    function publishPrice(
+        bytes32 identifier,
+        uint256 time,
+        bytes memory ancillaryData
+    ) public {
+        require(_getOracle().hasPrice(identifier, time, ancillaryData), "DVM has not resolved price");
+        int256 price = _getOracle().getPrice(identifier, time, ancillaryData);
+        // This implementation allows duplicate MessageSent events via _sendMessageToRoot. The child tunnel on the
```

Comment:
Important to read this design decision. This is based on feedback from Polygon team. The benefit of allowing duplicate calls to `_sendMessageToChild` is ensuring that we can resubmit any transactions that the bridge might miss. The downside is that anyone can call this method to spam the Polygon hosted tunnel. Thoughts?

Reply:
Perhaps we can impose a rate-limit within the contract.

Reply:
well seeing the price that is pulled here comes directly from the oracle there is not a huge amount of risk IMHO. what would you be able to do by spamming this? as the L2 has no rate limiting this should not be a problem on L1.

Reply:
If the L2 does not have any rate limiting and this does not call into and UMA DVM contracts other than the `hasPrice` and `viewPrice` view methods then I don't really see any downside in this.

Reply:
You would just be able to spam the Polygon-hosted relayers, which doesn't seem like a big risk for us.

Reply:
The only risk I could see is that they somehow blacklist _our_ contracts because of spam. But my guess is that they won't do that.